stl/inc/utility

```diff
@@ -644,6 +645,77 @@ template <size_t _Idx>
 inline constexpr in_place_index_t<_Idx> in_place_index{};
 #endif // _HAS_CXX17
 
+#if _HAS_CXX20
+
+template <typename _Ty>
+inline constexpr bool _Is_integral_not_bool_char = is_integral_v<_Ty>
+    && !is_same_v<_Ty, bool> && !is_same_v<_Ty, char>
+    && !is_same_v<_Ty, char8_t> && !is_same_v<_Ty, char16_t>
+    && !is_same_v<_Ty, char32_t> && !is_same_v<_Ty, wchar_t>;
+
+// SAFE INTEGRAL COMPARISONS
+template <typename _Ty1, typename _Ty2>
+constexpr bool in_range(const _Ty2 _Value) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>);
+    if constexpr (is_signed_v<_Ty1> && is_unsigned_v<_Ty2>) {
+        return _Value <= (numeric_limits<_Ty1>::max)();
+    } else if constexpr (is_unsigned_v<_Ty1> && is_signed_v<_Ty2>) {
+        return _Value >= 0 && _Value <= (numeric_limits<_Ty1>::max)();
+    } else {
+        return (numeric_limits<_Ty1>::min)() <= _Value && _Value <= (numeric_limits<_Ty1>::max)();
+    }
+}
+
+template <typename _Ty1, typename _Ty2>
+constexpr bool cmp_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>);
+    if constexpr (is_signed_v<_Ty1> && is_unsigned_v<_Ty2>) {
+        return _Left >= 0 || static_cast<make_unsigned_t<_Ty1>>(_Left) == _Right;
+    } else if constexpr (is_unsigned_v<_Ty1> && is_signed_v<_Ty2>) {
+        return _Right >= 0 && _Left == static_cast<make_unsigned_t<_Ty2>>(_Right);
+    } else {
+        return _Left == _Right;
+    }
+}
+
+template <typename _Ty1, typename _Ty2>
+constexpr bool cmp_not_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>);
+    return !cmp_equal(_Left, _Right);

```

Comment:
I think @BillyONeal cares about debug performance so I would guess you have to open-code it.

That said these comparison functions are so hairy that you might be lucky (Please wait for a maintainer to confirm / reject before you do any work)

Reply:
I tend to think that this is fine as-is; it's true that it pays one more function call for debug perf/codegen size, but I don't expect these to be truly hammered by applications, and any application that really does hammer this could be writing the comparisons directly (carefully). As you observed, there is a significant cost to replicating these complicated functions with significant mutations. We can always change this code later if we receive reported issues.

(My opinion matters the least here, as I am historically the least sensitive to debug perf concerns, although I am trying to reform :cat: )